themes/pure/pure.theme.bash

```diff
@@ -17,30 +19,30 @@ VIRTUALENV_THEME_PROMPT_SUFFIX=")"
 # export LS_COLORS='no=00:fi=00:di=01;34:ln=00;36:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=41;33;01:ex=00;32:*.cmd=00;32:*.exe=01;32:*.com=01;32:*.bat=01;32:*.btm=01;32:*.dll=01;32:*.tar=00;31:*.tbz=00;31:*.tgz=00;31:*.rpm=00;31:*.deb=00;31:*.arj=00;31:*.taz=00;31:*.lzh=00;31:*.lzma=00;31:*.zip=00;31:*.zoo=00;31:*.z=00;31:*.Z=00;31:*.gz=00;31:*.bz2=00;31:*.tb2=00;31:*.tz2=00;31:*.tbz2=00;31:*.avi=01;35:*.bmp=01;35:*.fli=01;35:*.gif=01;35:*.jpg=01;35:*.jpeg=01;35:*.mng=01;35:*.mov=01;35:*.mpg=01;35:*.pcx=01;35:*.pbm=01;35:*.pgm=01;35:*.png=01;35:*.ppm=01;35:*.tga=01;35:*.tif=01;35:*.xbm=01;35:*.xpm=01;35:*.dl=01;35:*.gl=01;35:*.wmv=01;35:*.aiff=00;32:*.au=00;32:*.mid=00;32:*.mp3=00;32:*.ogg=00;32:*.voc=00;32:*.wav=00;32:'
 
 scm_prompt() {
-    CHAR=$(scm_char)
```

Comment:
Why this is not declared as `local`? I grepped `\bCHAR\b` in `bash-it` directory and found that there many similar functions, but no other part references the variable `CHAR`:

<details><summary>[Result of <code>grep -C3 '\bCHAR\b' ...</code>]</summary>

```bash
./themes/binaryanomaly/binaryanomaly.theme.bash-32-}
./themes/binaryanomaly/binaryanomaly.theme.bash-33-
./themes/binaryanomaly/binaryanomaly.theme.bash-34-scm_prompt() {
./themes/binaryanomaly/binaryanomaly.theme.bash:35:     CHAR=$(scm_char)
./themes/binaryanomaly/binaryanomaly.theme.bash:36:     if [ "$CHAR" = "$SCM_NONE_CHAR" ]; then
./themes/binaryanomaly/binaryanomaly.theme.bash-37-             return
./themes/binaryanomaly/binaryanomaly.theme.bash-38-     else
./themes/binaryanomaly/binaryanomaly.theme.bash-39-             echo "[$(scm_char)$(scm_prompt_info)]"
--
./themes/brunton/brunton.theme.bash-16-}
./themes/brunton/brunton.theme.bash-17-
./themes/brunton/brunton.theme.bash-18-scm_prompt() {
./themes/brunton/brunton.theme.bash:19: CHAR=$(scm_char)
./themes/brunton/brunton.theme.bash:20: if [ "$CHAR" = "$SCM_NONE_CHAR" ]; then
./themes/brunton/brunton.theme.bash-21-         return
./themes/brunton/brunton.theme.bash-22- else
./themes/brunton/brunton.theme.bash-23-         echo " $(scm_char) (${white}$(scm_prompt_info)${normal})"
--
./themes/codeword/codeword.theme.bash-4-SCM_GIT_CHAR="${green}±${normal}"
./themes/codeword/codeword.theme.bash-5-
./themes/codeword/codeword.theme.bash-6-scm_prompt() {
./themes/codeword/codeword.theme.bash:7:    CHAR=$(scm_char)
./themes/codeword/codeword.theme.bash:8:    if [ $CHAR = $SCM_NONE_CHAR ]
./themes/codeword/codeword.theme.bash-9-        then
./themes/codeword/codeword.theme.bash-10-            return
./themes/codeword/codeword.theme.bash-11-        else
--
./themes/doubletime/doubletime.theme.bash-29-fi
./themes/doubletime/doubletime.theme.bash-30-
./themes/doubletime/doubletime.theme.bash-31-doubletime_scm_prompt() {
./themes/doubletime/doubletime.theme.bash:32:  CHAR=$(scm_char)
./themes/doubletime/doubletime.theme.bash:33:  if [ $CHAR = $SCM_NONE_CHAR ]; then
./themes/doubletime/doubletime.theme.bash-34-    return
./themes/doubletime/doubletime.theme.bash:35:  elif [ $CHAR = $SCM_GIT_CHAR ]; then
./themes/doubletime/doubletime.theme.bash-36-    echo "$(git_prompt_status)"
./themes/doubletime/doubletime.theme.bash-37-  else
./themes/doubletime/doubletime.theme.bash-38-    echo "[$(scm_prompt_info)]"
--
./themes/gallifrey/gallifrey.theme.bash-14-# export LS_COLORS='no=00:fi=00:di=01;34:ln=00;36:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=41;33;01:ex=00;32:*.cmd=00;
32:*.exe=01;32:*.com=01;32:*.bat=01;32:*.btm=01;32:*.dll=01;32:*.tar=00;31:*.tbz=00;31:*.tgz=00;31:*.rpm=00;31:*.deb=00;31:*.arj=00;31:*.taz=00;31:*.lzh=00;31:*.lzma=00;31:*.
zip=00;31:*.zoo=00;31:*.z=00;31:*.Z=00;31:*.gz=00;31:*.bz2=00;31:*.tb2=00;31:*.tz2=00;31:*.tbz2=00;31:*.avi=01;35:*.bmp=01;35:*.fli=01;35:*.gif=01;35:*.jpg=01;35:*.jpeg=01;35
:*.mng=01;35:*.mov=01;35:*.mpg=01;35:*.pcx=01;35:*.pbm=01;35:*.pgm=01;35:*.png=01;35:*.ppm=01;35:*.tga=01;35:*.tif=01;35:*.xbm=01;35:*.xpm=01;35:*.dl=01;35:*.gl=01;35:*.wmv=0
1;35:*.aiff=00;32:*.au=00;32:*.mid=00;32:*.mp3=00;32:*.ogg=00;32:*.voc=00;32:*.wav=00;32:'
./themes/gallifrey/gallifrey.theme.bash-15-
./themes/gallifrey/gallifrey.theme.bash-16-scm_prompt() {
./themes/gallifrey/gallifrey.theme.bash:17:    CHAR=$(scm_char)
./themes/gallifrey/gallifrey.theme.bash:18:    if [ $CHAR = $SCM_NONE_CHAR ]
./themes/gallifrey/gallifrey.theme.bash-19-        then
./themes/gallifrey/gallifrey.theme.bash-20-            return
./themes/gallifrey/gallifrey.theme.bash-21-        else
--
./themes/mairan/mairan.theme.bash-70-}
./themes/mairan/mairan.theme.bash-71-
./themes/mairan/mairan.theme.bash-72-modern_scm_prompt() {
./themes/mairan/mairan.theme.bash:73:        CHAR=$(scm_char)
./themes/mairan/mairan.theme.bash:74:        if [ $CHAR = $SCM_NONE_CHAR ]
./themes/mairan/mairan.theme.bash-75-        then
./themes/mairan/mairan.theme.bash-76-                return
./themes/mairan/mairan.theme.bash-77-        else
--
./themes/modern-t/modern-t.theme.bash-33-}
./themes/modern-t/modern-t.theme.bash-34-
./themes/modern-t/modern-t.theme.bash-35-modern_scm_prompt() {
./themes/modern-t/modern-t.theme.bash:36:       CHAR=$(scm_char)
./themes/modern-t/modern-t.theme.bash:37:       if [ $CHAR = $SCM_NONE_CHAR ]
./themes/modern-t/modern-t.theme.bash-38-       then
./themes/modern-t/modern-t.theme.bash-39-               return
./themes/modern-t/modern-t.theme.bash-40-       else
--
./themes/modern-time/modern-time.theme.bash-29-}
./themes/modern-time/modern-time.theme.bash-30-
./themes/modern-time/modern-time.theme.bash-31-modern_scm_prompt() {
./themes/modern-time/modern-time.theme.bash:32: CHAR=$(scm_char)
./themes/modern-time/modern-time.theme.bash:33: if [ $CHAR = $SCM_NONE_CHAR ]
./themes/modern-time/modern-time.theme.bash-34- then
./themes/modern-time/modern-time.theme.bash-35-         return
./themes/modern-time/modern-time.theme.bash-36- else
--
./themes/modern/modern.theme.bash-27-}
./themes/modern/modern.theme.bash-28-
./themes/modern/modern.theme.bash-29-modern_scm_prompt() {
./themes/modern/modern.theme.bash:30:   CHAR=$(scm_char)
./themes/modern/modern.theme.bash:31:   if [ "$CHAR" = "$SCM_NONE_CHAR" ]; then
./themes/modern/modern.theme.bash-32-           return
./themes/modern/modern.theme.bash-33-   else
./themes/modern/modern.theme.bash-34-           echo "[$(scm_char)][$(scm_prompt_info)]"
--
./themes/pure/pure.theme.bash-17-# export LS_COLORS='no=00:fi=00:di=01;34:ln=00;36:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=41;33;01:ex=00;32:*.cmd=00;32:*.exe=0
1;32:*.com=01;32:*.bat=01;32:*.btm=01;32:*.dll=01;32:*.tar=00;31:*.tbz=00;31:*.tgz=00;31:*.rpm=00;31:*.deb=00;31:*.arj=00;31:*.taz=00;31:*.lzh=00;31:*.lzma=00;31:*.zip=00;31:
*.zoo=00;31:*.z=00;31:*.Z=00;31:*.gz=00;31:*.bz2=00;31:*.tb2=00;31:*.tz2=00;31:*.tbz2=00;31:*.avi=01;35:*.bmp=01;35:*.fli=01;35:*.gif=01;35:*.jpg=01;35:*.jpeg=01;35:*.mng=01;
35:*.mov=01;35:*.mpg=01;35:*.pcx=01;35:*.pbm=01;35:*.pgm=01;35:*.png=01;35:*.ppm=01;35:*.tga=01;35:*.tif=01;35:*.xbm=01;35:*.xpm=01;35:*.dl=01;35:*.gl=01;35:*.wmv=01;35:*.aif
f=00;32:*.au=00;32:*.mid=00;32:*.mp3=00;32:*.ogg=00;32:*.voc=00;32:*.wav=00;32:'
./themes/pure/pure.theme.bash-18-
./themes/pure/pure.theme.bash-19-scm_prompt() {
./themes/pure/pure.theme.bash:20:    CHAR=$(scm_char)
./themes/pure/pure.theme.bash:21:    if [ $CHAR = $SCM_NONE_CHAR ]
./themes/pure/pure.theme.bash-22-        then
./themes/pure/pure.theme.bash-23-            return
./themes/pure/pure.theme.bash-24-        else
--
./themes/ramses/ramses.theme.bash-37-}
./themes/ramses/ramses.theme.bash-38-
./themes/ramses/ramses.theme.bash-39-modern_scm_prompt() {
./themes/ramses/ramses.theme.bash:40:        CHAR=$(scm_char)
./themes/ramses/ramses.theme.bash:41:        if [ $CHAR = $SCM_NONE_CHAR ]
./themes/ramses/ramses.theme.bash-42-        then
./themes/ramses/ramses.theme.bash-43-                return
./themes/ramses/ramses.theme.bash-44-        else
--
./themes/rjorgenson/rjorgenson.theme.bash-58-}
./themes/rjorgenson/rjorgenson.theme.bash-59-
./themes/rjorgenson/rjorgenson.theme.bash-60-modern_scm_prompt() {
./themes/rjorgenson/rjorgenson.theme.bash:61:        CHAR=$(scm_char)
./themes/rjorgenson/rjorgenson.theme.bash:62:        if [ $CHAR = $SCM_NONE_CHAR ]
./themes/rjorgenson/rjorgenson.theme.bash-63-        then
./themes/rjorgenson/rjorgenson.theme.bash-64-                return
./themes/rjorgenson/rjorgenson.theme.bash-65-        else
./themes/rjorgenson/rjorgenson.theme.bash:66:                echo "${BRACKET_COLOR}[${CHAR}${BRACKET_COLOR}][${STRING_COLOR}$(scm_prompt_info)${BRACKET_COLOR}]$normal"
./themes/rjorgenson/rjorgenson.theme.bash-67-        fi
./themes/rjorgenson/rjorgenson.theme.bash-68-}
./themes/rjorgenson/rjorgenson.theme.bash-69-
--
./themes/slick/slick.theme.bash-44-}
./themes/slick/slick.theme.bash-45-
./themes/slick/slick.theme.bash-46-modern_scm_prompt() {
./themes/slick/slick.theme.bash:47:        CHAR=$(scm_char)
./themes/slick/slick.theme.bash:48:        if [ $CHAR = $SCM_NONE_CHAR ]
./themes/slick/slick.theme.bash-49-        then
./themes/slick/slick.theme.bash-50-                return
./themes/slick/slick.theme.bash-51-        else
--
./themes/zork/zork.theme.bash-37-}
./themes/zork/zork.theme.bash-38-
./themes/zork/zork.theme.bash-39-modern_scm_prompt() {
./themes/zork/zork.theme.bash:40:        CHAR=$(scm_char)
./themes/zork/zork.theme.bash:41:        if [ $CHAR = $SCM_NONE_CHAR ]
./themes/zork/zork.theme.bash-42-        then
./themes/zork/zork.theme.bash-43-                return
./themes/zork/zork.theme.bash-44-        else
```
</details>

I think all of these `CHAR`s should be declared with `local`.

Also, there are codes similar to the one caused the problem in `pure.theme.bash` in many other themes. Shouldn't they also be updated?

Reply:
You are very correct. I think I will split this function to a centralised location, instead of redefining it everywhere like now

Reply:
Updated by moving `scm_prompt` to a central location, tested all of the touched themes to see that they look the same :smile: